\chapter{The Limited Proteolysis module}
\label{chap:limprot}

The Limited Proteolysis module is designed to post-process the results from an enzymatic
digestion performed in two steps. The first step is assumed to be a limited proteolysis
in which a large protein is split in smaller fragments. The fragments are then separated
using a SDS-PAGE electrophoresis. Finally, selected bands from the gel are submitted to
a full enzymatic digestion and the generated peptides are analyzed using mass spectrometry.

The main objective of the module is to identify the protein fragments generated in
the initial limited proteolysis from the peptides found in the MS analyzed gel spots.
This is achieved by performing an equivalence test\cite{Limentani2005} between the
peptides in the selected gel spots and a control spot containing the full length
target protein. In this way, peptides leaked from one gel spot to another can be
eliminated. Several replicates of the experiment are expected.

\section{Definitions}
\label{sec:limprotDefinitions}

Before explaining in detail the interface of the module and how does the module
work, let's make clear the meaning of some terms that will be used in the following
paragraphs.

\phantomsection
\begin{itemize}
    \item \textit{Recombinant protein}: actual amino acid sequence used in the mass
    spectrometry experiments. It may be identical to the native sequence of the Target
    protein under study or not.
    \item \textit{Native protein}: full amino acid sequence expressed in wild type cells.
    \item \textit{Detected peptide}: any peptide detected in any of the mass spectrometry
    experiments including the control experiments.
    \item \textit{Relevant peptide}: a detected peptide with a Score value above
    a user defined threshold, see page \pageref{par:limprotScoreValue}.
    \item \textit{Filtered peptide}: a relevant peptide with equivalent intensities
    in the control and a given gel spot at the chosen significance level.\label{par:limprotFP}
    \item \textit{Fragment}: group of filtered peptides with no gaps when their
    sequences are aligned to the sequence of the recombinant/native protein.
\end{itemize}

For example, there are three fragments in the alignment shown below. The first fragment
is formed by sequences \numrange{1}{3} since there is no gap in the sequence MKKTAIAIAVAL.
SEQ\num{4} forms the second fragment because there is a gap between the last residue
in SEQ\num{3} and the first residue in SEQ\num{4} and another gap between the last
residue in SEQ\num{4} and the first residue in SEQ\num{5}. For the same reason
SEQ\num{5} forms the third fragment.

\begin{texshade}{./TeX_files/test.fasta}
    \residuesperline*{50}
    \shadingmode[kenny]{functional}
    \hideconsensus
\end{texshade}

\section{The input files}

The Limited Proteolysis module requires a Data file containing the detected peptides
and a sequence file containing the amino acid sequence of the recombinant protein
used in the study. Both files must follow the guidelines specified in \autoref{sec:dataFile}.
In short, the Data file must have a tabular format with tab separated columns and
the name of the columns are expected as first row. The Sequence file is expected
to contain at least one sequence and to be FASTA formatted. If more than one sequence
is found in the Sequence file the first sequence will be taken as the sequence of
the Recombinant protein and the second sequence will be taken as the sequence of
the Native protein. All other sequences are discarded.

\section{The interface}

The tab of the Limited Proteolysis module is divided in two regions (\autoref{fig:limprotTab}).

\begin{figure}[h]
    \centering
    \includegraphics[width=0.7\textwidth]{./IMAGES/MOD-LIMPROT/limprot-mod.jpg}
    \caption[The Limited Proteolysis module tab]{\textbf{The Limited Proteolysis
    module tab.} This tab allows users to perform the analysis of the results obtained
    during a two steps enzymatic proteolysis experiment where the products from the
    first limited digestions are separated using SDS-PAGE electrophoresis.}
    \label{fig:limprotTab}
    \vspace{-5pt}
\end{figure}

The Data File Content region holds only a table to show the name of the columns in
the selected Data File. The table will be automatically filled after selecting the
file.

The Configuration Options region contains all the fields needed to configure and
run the analysis.

Section Files contains three buttons and a text field. Here users select the input
and output files for the analysis.

\num{1}. The button UMSAP allows users to browse the file system to select the location
and name of the .umsap file. When selecting an already existing .umsap file the operating
system will ask if it is ok to replace the file, the answer can be yes since UMSAP
will never overwrite or replace an .umsap file, instead the new analysis will be
added to the already existing file. Only .umsap files can be selected here.

\num{2}. The button Data allows users to browse the file system to select the input
data file that will be used for the analysis. The Data file is expected to be a
plain text file with tab separated columns and the name of the columns in the first
row of the file. In addition, columns to be analyzed must contain only numbers and
must be of the same length. Only .txt files can be selected here.

\num{3}. The button Sequences allows users to browse the file system to select the
FASTA file containing the sequence of the Recombinant protein and the Native protein.
The FASTA file must contain at least one sequence.

\num{4}. The text field Analysis ID allows users to provide an ID for the analysis
to be run. The date and time of the analysis will be automatically added to the
beginning of the name.

Section Data Preparation contains four dropdown boxes. Here users select how the data
in the Data file should be prepared before starting the analysis.

\num{1}. The dropdown Treat \num{0}s as missing values allows user to define how
to handle \num{0} values present in the Data file.

\num{2}. The dropdown Transformation allows user to select the Transformation method
to be applied to the data.

\num{3}. The dropdown Normalization allows users to select the Normalization method
to be applied to the data.

\num{4}. The dropdown Imputation allows user to select the Imputation method used
to replace missing values in the data.

Section User-defined values contains seven text fields and one dropdown box. Here
users configure the Limited Proteolysis to be run.

\phantomsection
\num{1}. The text field Target Protein\label{par:limprotTargetProtein} allows users
to specify the protein of interest. Users may type here any unique protein identifier
present in the Data file. The search for the Target Protein is case-sensitive, meaning
that eFeB is not the same as efeb.

\phantomsection
\num{2}. The text field Score Value\label{par:limprotScoreValue} allows users to
define a threshold value above which the detected peptides will be considered as
relevant. The Score Value is an indicator of how reliable was the detection of the peptide
during the MS experiments. The value given to UMSAP depends on the program generating
the Data file. Only one real number equal or greater than zero will be accepted here.
A value of zero means all detected peptides belonging to the Target Protein will
be treated as relevant peptides.

\num{3}. The dropdown Samples allows users to specify whether samples are independent
or paired. For example, samples are paired when the same Petri dish is used for the
control and experiments.

\numrange[range-phrase = --]{4}{8}. The parameters $\alpha$, $\beta$, $\gamma$,
$\Theta$ and $\Theta$max are used to configure the equivalence test\cite{Limentani2005}
performed to identify peptides in the selected gel spots with equivalent intensity
values to the control spots (\autoref{sec:limprotEquivalenceTest}). $\alpha$, $\beta$ and 
$\gamma$ must be between \num{0} and \num{1}. The value of $\Theta$ is optional. If
left blank UMSAP will calculate a value for each peptide based on the intensity values
found in the Data file. If given then the given value will be used for each peptide.
$\Theta$max is the maximum value to consider the intensity values in the gel spot
and control as equivalent.

Section Column numbers contains four text fields. Here, users provide the column
numbers in the Data file from where UMSAP will get the information needed to perform
the Limited Proteolysis analysis. All columns specified in this section must be present
in the Data file. Column numbers start at \num{0}. The column numbers are shown in
the table of Region Data File Content after the Data file is selected.

\num{1}. The text field Sequences allows users to specify the column in the Data
file containing the sequences of the peptides identified in the MS experiments.
Only one integer number equal or greater than zero will be accepted here.

\num{2}. The text field Detected Proteins allows users to specify the column in
the Data file containing the unique protein identifier for the proteins detected
in the MS experiments. It is in this column where the program will look for the
Target Protein value given in Section User-defined values. It is important that
in this column the Target Protein value is used to refer to only one protein. Only
one integer number equal or greater than zero will be accepted here.

\num{3}. The text field Score allows users to specify the column in the Data file
containing the Score values. It is in this column where the program will look for
the values to be compared against the Score threshold given in Section User-defined
values.

\phantomsection
\num{4}. \label{par:limprotResultControl}The text field Results - Control experiments
allows users to specify the columns in the Data file containing the results of the
control and experiments. The button Type Values call a helper window
(\autoref{fig:limprotResControlWindow}) where users can type the information needed. 

\begin{figure}[h]
    \centering
    \includegraphics[width=0.7\textwidth]{./IMAGES/MOD-LIMPROT/limprot-rescontrol.jpg}
    \caption[The Result - Control experiments helper window for the Limited Proteolysis module]{\textbf{The Result - Control experiments helper window for the Limited Proteolysis module.} This window allows users to specify the column numbers in the Data file containing the MS results for the selected gel spots.} 
    \label{fig:limprotResControlWindow}
    \vspace{-5pt} 	
\end{figure}

The helper window is divided in two Regions. Region Data File Content will show the 
column numbers and names of the columns present in the selected Data file. Region
Configuration Options has two sections. The upper section allows defining the number
of bands and lanes of interest in the gel as well as the label for lanes, bands and
control spot. The button Setup Fields creates the corresponding text fields in the
bottom section to type the column numbers. Each text field should contain the column
numbers with the MS results for the given gel spot. The values for the text fields
should be positive integer numbers or a range of integers, e.g. 
\numrange[range-phrase=--]{60}{62} or left blank for empty gel spots. Selected rows
in the table can be copied (Cmd+C) and then pasted (Cmd+V) in the text fields. 
Duplicate column numbers are not allowed. 

\section{The analysis}
\label{sec:limprotEquivalenceTest}
First, UMSAP will check the validity of the user provided input. Then, the Data file is processed as follow. All rows in the Data file containing peptides that do not belong to the Target protein are removed. Then, all rows containing peptides from the Target protein but with Score values lower than the user defined Score threshold are removed. These steps leave only relevant peptides, this means, peptides with a Score value higher than the user defined threshold that belong to the Target protein. For each one of these relevant peptides the equivalence test is performed \cite{Limentani2005}.

The implementation of the equivalence test is based on the following equations:
\begin{equation}
    \label{eq:limprotDeviationUpperLimit}
    s^* = s\sqrt{\frac{n-1}{\chi^2_{(\gamma, n-1)}}}
\end{equation}
\begin{equation}
\label{eq:limprotAcceptanceCriterion}
\Theta = \delta + s^*\left[t_{(1-\alpha, 2n-2)} + t_{(1-\beta/2, 2n-2)}\right]\sqrt{\frac{2}{n}}
\end{equation}
\begin{equation}
\label{eq:limprotEquivalenceTest}
(\bar{y}_1 - \bar{y}_2) \pm t_{(1-\alpha, n_1+n_2-2)} \cdot \sqrt{s^2_p\left(\frac{1}{n_1}+\frac{1}{n_2}\right)}
\end{equation}

where $s^*$ is an estimate of the upper confidence limit of the standard deviation, $\chi^2_{(\gamma, n-1)}$ is the ($100\gamma$)th percentile of the chi-squared distribution with $n-1$ degrees of freedom, $\Theta$ is the acceptance criterion, $\delta$ is the absolute value of the true difference between the group's mean values, $t$ is the Student's $t$ value, $\bar{y}$ is the measurement mean and $s_p$ is the pooled standard deviation of the measurements calculated with:
\begin{equation}
\label{eq:poolStDev}
s_p =  \sqrt{\frac{(n_1-1)s^2_1+(n_2-1)s^2_2}{n_1+n_2-2}}
\end{equation}
$\alpha$, $\beta$, $\gamma$ and $\Theta$ are the parameters defined in section Values of Region \num{2} of the main window of the module.

In essence, for each relevant peptide, the control experiments are used to estimate the upper confidence limit for the standard deviation using \autoref{eq:limprotDeviationUpperLimit} and then the acceptance criterion is calculated with \autoref{eq:limprotAcceptanceCriterion}. Finally, the confidence interval for the mean difference for the gel spot and the control is calculated with \autoref{eq:limprotEquivalenceTest} and compare to $\Theta$. Peptides with equivalent mean intensity in at least one gel spot and the control are retained while not equivalent peptides are discarded. 

If the value of $\Theta$ is given in Region \num{2} of the module's interface then only the confidence interval for the mean difference is calculated and the value is directly compared to the given $\Theta$ value. The maximum possible $\Theta$ value must always be provided. The reason for this is that when only a few replicates of the experiments are performed the calculated $\Theta$ value may be to large and then the equivalence test is not able to detect the peptides with intensity values in the experiments lower than in the control.

If the sequence of the native protein is given the module performs a sequence alignment between the native and recombinant sequences. The alignment allows UMSAP to translate the results obtained with the residue numbers of the recombinant protein to the residue numbers of the native protein. This is done to facilitate future comparison of results between different recombinant proteins of the same native protein. However, when analyzing the results of the alignment the module assumes that the recombinant and native sequences differs only in the N and C-terminal tags while the sequence between the tags is identical. If this is not the case, e.g. there are point mutations or insertion/deletion in the sequence of the recombinant protein no native sequence file should be given to UMSAP.

After the filtered peptide (FP) are identified the modules creates the output files.

\section{The output files}

All the output generated by the Limited Proteolysis module will be contained in a LimProt folder created inside the selected Output folder. If the Output folder field in section \textit{Files} of Region \num{2} of the interface is left empty, then the LimProt folder will be created in the directory containing the Data file. If the selected Output folder already contains a LimProt folder, then the current date and time to the second will be added to the name in order to avoid overwriting files from previous analyses. By default the LimProt folder will contain two files with extensions .limprot and .uscr and a Data{\_}Steps folder. The name of these files is provided with the Output name field in section \textit{Files} of Region \num{2} of the interface. Depending on the user provided input extra folders and files will be created inside the LimProt folder (\autoref{fig:limprotOutFolder}). For the rest of this chapter we will assume that the user provided name for the Output folder was \textit{t}, the Output name was myLimTest, the Target protein was \textit{Mis18alpha} and all optional analyses were performed.

Information regarding the content and use of the .uscr file can be found in \autoref{subsec:utilUscrFile}.

If the parameter Sequence length is different than NA, then the file myLimTest.seq.pdf will be created. The file contains the sequence of the Target protein with the sequence of the FP highlighted. More details are given in \autoref{subsec:utilSeqHigh}.

If the parameter Columns to extract is different than NA, then the folder Data will be created. The folder will contain several files that are shorter versions of the Data file. These files will contain only information regarding the Target protein. More details are given in \autoref{subsec:utilShortDF}.

The folder Data\_Steps contains a step by step account of all the calculations performed so users can check the accuracy of the calculation or perform further analysis. The files inside Data\_Steps are plain text file with tab separated columns. The first line contains the name of the columns in the file.

A file containing a list of FP can also be generated as indicated in \autoref{subsec:utilExpData}.

\begin{figure}[h]
    \centering
    \includegraphics[width=0.35\textwidth]{./IMAGES/MOD-LIMPROT/limprot-files.jpg}	    
    \caption[The structure of the Output folder from the Limited Proteolysis module]{\textbf{The structure of the Output folder from the Limited Proteolysis module.} The folder Data and the file myLimTest.seq.pdf will be created only if requested.} 
    \label{fig:limprotOutFolder}
    \vspace{-5pt} 	
\end{figure}

\section{Visualizing the output files}

The main output of the module is the .limprot file. This file contains the list of FP,  all parameters values and all the information needed to visualize the results with UMSAP. After creating the file at the end of the analysis, the Limited Proteolysis module will automatically load the file and create a windows to display the results, see \autoref{fig:limprotResultsWindow}.

\begin{figure}[h]
    \centering
    \includegraphics[width=0.8\textwidth]{./IMAGES/MOD-LIMPROT/limprot-frag.jpg}	    
    \caption[The Gel Analysis window]{\textbf{The Gel Analysis window.} Users can performed here the analysis of the fragments obtained in the limited proteolysis experiments.} 
    \label{fig:limprotResultsWindow}
    \vspace{-5pt} 	
\end{figure}

The Gel analysis window is divided in four Regions.

Region \num{1} contains a list of all FP contained in the .limprot file being shown. The search box at the bottom allows to search for a sequence in the list of FP. 

Region \num{2} contains a representation of the analyzed gel. Here, each gel spot is represented with square. When a square is not shown this means that the corresponding gel spot was not analyzed. Empty squares represent gel spot where no peptide from the Target protein was detected with intensity values equivalent to the controls. The rest of the square will be colored according to the band they belong to or the lane. There are two selection modes available for Region \num{2}. In the Lane selection mode selecting one of the Gel spot will also select the entire lane containing the selected gel spot. In this mode the gel spot will be colored according to the band they belong to. The selected lane will be highlighted with a red rectangle. The right mouse button, the Tools menu or the keyboard shortcut Ctrl/Cmd+L can be used to change the selection mode. In the Band selection mode, selecting a gel spot will highlight the band containing the gel spot and the gel spots are colored by lane. Selecting a band or a lane in Region \num{2} will display information about the band/lane in Regions \num{3} and \num{4}.

Region \num{3} will display a graphical representation of the fragments found in each gel spot for the selected band or lane. The first fragment in this region represent the full length of the recombinant sequence of the Target protein. Here, the central red section represents the sequence in the recombinant protein that is identical to the native protein sequence while gray sections represent the sequences in the recombinant protein that are different to the native protein sequence.  If the sequence of the native protein was not given then the fragment is shown in gray. The fragments are color coded using the same colors of the band/lane they belong to.

Selecting a peptide from the list box in Region \num{1} will highlight with a blue border the gel spots in Region \num{2} where the peptide is found. If a lane/band in Region \num{2} is already selected, then the fragments shown in Region \num{3} that contains the selected peptide in the list box will also be highlighted with a blue border.

Region \num{4} will show information about the selected lane/band or gel spot in Region \num{2} or the selected fragment in Region \num{3}. The displayed information for a selected band/lane includes the number of non-empty lanes/bands, the number of fragments identified in each non-empty gel spot in the band/lane and the protein regions identified. Selecting a gel spot will display this information only for the gel spot. Selecting a fragment in Region \num{3} will display in Region \num{4} the following information: number of cleavage sites and fragments, first and last residue number for the selected fragment and a sequence alignment of all peptides forming the fragment.  

\subsection{The Tools menu}
\label{subsec:limprotToolsMenu}

The Tools menu for the window allows changing the selection mode in Region \num{2}, export the data shown in the window (see \autoref{subsec:utilExpData}), save an image of Region \num{2} or \num{3} and to reset the view of the window.

The file containing the exported list of FP will have a tabular format with tab delimited columns. Each row of the file will contain a FP. The columns in the file contain information about the N and C residue numbers, the sequence and Score of the FP. In addition, the results of the equivalence test for each gel spot is given as 0 or 1 value, with 0 meaning not equivalent. The file will have a .txt extension and can be viewed with a simple text editor or Excel.