\chapter{Proteome Profiling}
\label{chap:protprof}

The Proteome Profiling modules is designed to identify differentially expressed
protein under various experimental conditions. A typical example is to compare the
effect of two substance over protein expression using whole cell lysates.

\section{Definitions}

Before explaining in detail the interface of the module and how does the module work, 
let's make clear the meaning of some terms that will be used in the following paragraphs.

\phantomsection
\begin{itemize}
	\item \textit{Detected protein}: any protein detected in any of the mass spectrometry
    experiments including the control experiments.
	\item \textit{Relevant proteins}: a detected protein with a Score value above
    a user-defined thresholds (page \pageref{par:protprofScoreValue}).
\end{itemize}

\section{The input files}

The Proteome Profiling module requires only one input file. This Data file must follow
the guidelines specified in \autoref{sec:dataFile}. In short, the Data file must
have a tabular format with tab separated columns and the name of the columns are
expected as first row. All columns given as input in section Column numbers of Region
Configuration Options of the interface (\autoref{fig:protprofTab}) must be present
in the Data file.

\section{The interface}

The tab of the Proteome Profiling module is divided in two regions (\autoref{fig:protprofTab}).

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/MOD-PROTPROF/protprof-mod.jpg}
	\caption[The Proteome Profiling module tab]{\textbf{The Proteome Profiling module
    tab.} This tab allows users to perform a proteome profiling analysis.}
	\label{fig:protprofTab}
	\vspace{-5pt}
\end{figure}

The Data File Content region holds only a table to show the name of the columns in
the selected Data File. The table will be automatically filled after selecting the
file. Selected rows in the table can be copied (Cmd+C) and pasted (Cmd+V) to the
text fields in region Configuration Options.

The Configuration Options region contains all the fields needed to configure and
run the analysis.

Section Files contains two buttons and a text field. Here users select the input
and output files for the analysis.

\num{1}. The button UMSAP allows users to browse the file system to select the location
and name of the .umsap file. When selecting an already existing .umsap file the operating
system will ask if it is ok to replace the file, the answer can be yes since UMSAP
will never overwrite or replace an .umsap file, instead the new analysis will be
added to the already existing file. Only .umsap files can be selected here.

\num{2}. The button Data allows users to browse the file system to select the input
data file that will be used for the analysis. The Data file is expected to be a
plain text file with tab separated columns and the name of the columns in the first
row of the file. In addition, columns to be analyzed must contain only numbers and
must be of the same length. Only .txt files can be selected here.

\num{3}. The text field Analysis ID allows users to provide an ID for the analysis
to be run. The date and time of the analysis will be automatically added to the
beginning of the name.

Section Data Preparation contains four dropdown boxes. Here users select how the data
in the Data file should be prepared before starting the analysis.

\num{1}. The dropdown Treat \num{0}s as missing values allows user to define how
to handle \num{0} values present in the Data file.

\num{2}. The dropdown Transformation allows user to select the Transformation method
to be applied to the data.

\num{3}. The dropdown Normalization allows users to select the Normalization method
to be applied to the data.

\num{4}. The dropdown Imputation allows user to select the Imputation method used
to replace missing values in the data.

Section User-defined values contains two text fields and three dropdown box. Here
users configure the Proteome Profiling analysis to be run.

\phantomsection
\num{1}. The text field Score Value\label{par:protprofScoreValue} allows users to
define a threshold value above which the detected proteins will be considered as
relevant. The Score value is an indicator of how reliable was the detection of
the protein during the MS experiments. The value given to UMSAP depends on the program
generating the Data file. Only one real number equal or greater than zero will be
accepted as a valid input here. A value of zero means all detected proteins will be
treated as relevant proteins.

\num{2}. The dropdown Samples allows users to specify whether samples are independent
or paired. For example, samples are paired when the same Petri dish is used for the
control and experiments.

\num{3}. The text field \alpha level allows users to define the significance level
used for the analysis. Only a number between \num{0} and \num{1} will be accepted
here.

\num{4}. The dropdown Intensities allows user to specify whether the intensity values
in the Data file represent absolute or a ratio of intensities.

\num{5}. The dropdown P Correction allows user to select the correction method for
the p values calculated during the analysis.

Section Column numbers contains five text fields. Here, users provide the column
numbers in the Data file from where UMSAP will get the information needed to perform
the analysis of the module. All columns specified in this section must be present
in the Data file. Column numbers start at \num{0}. The column numbers are shown in
the table of Region Data File Content after the Data file is selected.

\num{1}. The text field Detected Proteins allows users to specify the column in
the Data file containing the protein identifiers found in the Data file. Only one
integer number equal or greater than zero will be accepted here. 

\num{2}. The text field Gene Names allows users to specify the column in the Data
file containing the gene names of the proteins found during the MS experiments.
Only one integer number equal or greater than zero will be accepted here. 

\num{3}. The text field Score allows users to specify the column in the Data file
containing the Score values. It is in this column where the program will look for
the values to be compared against the Score threshold given in section User-defined
values. Only one integer number equal or greater than zero will be accepted here. 

\num{4}. The text field Exclude proteins allows users to specify several columns
in the Data file. Proteins found in these columns will be excluded form the analysis.
The module assumes that these columns contains numeric values and values greater
than zero indicate that the respective protein must be eliminated from the analysis.
Only integer numbers equal or greater than zero will be accepted here. If left empty
all proteins will be considered during the analysis. 

\num{5}. \label{par:protprofResultControl} The text field Results - Control experiments
allows users to specify the columns in the Data file containing the results of the
experiments. The button Type Values calls a helper window (\autoref{fig:protprofResControlWindow})
where users can type the information needed. Duplicate column numbers are not allowed here.

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/MOD-PROTPROF/protprof-rescontrol.jpg}
	\caption[The Result - Control experiments helper window for the Proteome Profiling module]{\textbf{The Result - Control experiments helper window for the Proteome Profiling module.} This window allows users to specify the column numbers in the Data file containing the MS results for the selected conditions, relevant points and control experiments.} 
	\label{fig:protprofResControlWindow}
	\vspace{-5pt} 	
\end{figure}

The helper window is divided in two Regions. Region Data File Content will show the 
column numbers and names of the columns present in the selected Data file. Region
Configuration Options has two sections. The upper section allows defining the number
of conditions and relevant points analyzed, to define the kind of control experiment
used as well as the label for conditions, relevant points and control experiment.
The button Setup Fields allows creating the matrix of text fields in the bottom section
where users type the column numbers. Each text field should contain the column
numbers with the MS results for the given experiment. The values for the text fields
should be positive integer numbers or a range of integers, e.g. 
\numrange[range-phrase=--]{60}{62} or left blank. Selected rows in the table can
be copied (Cmd+C) and then pasted (Cmd+V) in the text fields. Duplicate column numbers
are not allowed. 

\section{The analysis}
\label{sec:protprofTTest}

First, UMSAP will check the validity of the user provided input. In particular, all experiments need to have the same number of replicates as the respective control. Then, the Data file is processed as follow. All proteins found in the Exclude proteins columns are discarded. Proteins that were not identified in all conditions are discarded. Finally, all proteins with a Score value lower than the defined threshold are removed. The intensity values of the remaining proteins are normalized and then a median correction is applied to each experiment if Median correction was selected in section Values. With the resulting intensity values the fold change for each protein and experiment as well as the intensity ratios with respect to the control experiments are calculated and two different analysis are performed.

The fold change is calculated as: 

\begin{equation}
\label{eq:protprofFC}
FC = ave(I_{C, RP}) / ave(I_{Control})
\end{equation}

The first analysis is a t-test to determine if each experiment is significantly different to the corresponding control. The second analysis is a t-test, or ANOVA test if more than two conditions are studied, to determine if the values for the studied conditions are significantly different for each selected relevant point.

Finally, the corrected p values are calculated.

\section{The output files}

The Proteome Profiling module generates up to three files and a folder named Data\_Steps, see \autoref{fig:protprofOutFolder}. The folder Data\_Steps contains a step by step account of all the calculations performed so users can check the accuracy of the calculation or perform further analysis. The files inside Data\_Steps are plain text file with tab separated columns. The first line contains the name of the columns in the file.  

\begin{figure}[h]
	\centering
	\includegraphics[width=0.35\textwidth]{./IMAGES/MOD-PROTPROF/protprof-files.jpg}	    
	\caption[The structure of the Output folder from the Proteome Profiling module]{\textbf{The structure of the Output folder from the Proteome profiling module.} The file protprof.txt will be created only if requested.} 
	\label{fig:protprofOutFolder}
	\vspace{-5pt} 	
\end{figure}

The three files created have extension .txt, .uscr and .protprof. The file with extension .txt contains all the lines in the Data files but only the columns specified with the parameter Columns to extract in section Column numbers of Region \num{2} of the Proteome Profiling module. This file is generated only if the value of the parameter Columns to extract is not NA. 

The file with extension .uscr contains all the input given by the user so a new analysis may be performed without typing all the option values again, see \autoref{subsec:utilUscrFile}. 

The file with extension .protprof is the main output of the module. This file contains all the results and can be used to generate the graphical representation of the results.    

\section{Visualizing the output files}

After creating the .protprof at the end of the analysis, the Proteome Profiling module will automatically load the file and create a windows to display the results, see \autoref{fig:protprofResultsWindow}. This window is divided in four Regions. 

\begin{figure}[h]
	\centering
	\includegraphics[width=0.8\textwidth]{./IMAGES/MOD-PROTPROF/protprof-frag.jpg}	    
	\caption[The Proteome Profiling analysis window]{\textbf{The Proteome Profiling analysis window.} Users can performed here the analysis of the proteome profiling.} 
	\label{fig:protprofResultsWindow}
	\vspace{-5pt} 	
\end{figure}

Region \num{1} contains a list of all protein IDs and Gene names contained in the .protprof file being shown. The search box at the bottom allows to search for a protein in the list. Selecting a protein in the list will highlight the protein in Region \num{2} and display information about it in Regions \num{3} and \num{4}.

Region \num{2} contains a volcano plot showing the results for the t-test comparing the condition (C), relevant point (RP) to the corresponding control. The volcano plot has a horizontal line indicating the chosen significance level. In addition, the points in the plot can be colored by Z-score allowing to quickly identified the top up (blue) or down (red) regulated proteins. In \autoref{fig:protprofResultsWindow}, the top \SI{10}{\percent} up and down regulated proteins are colored and $\alpha$ is set to 0.05. Selecting a protein in the plot will highlight the selected protein in Region \num{1} and display information about it in Regions \num{3} and \num{4}. See \autoref{sec:protprofTools} for more options.

Region \num{3} contains a plot of $log_2[FC]$ vs Relevant points. The plot allows to see the behavior of the FC along the relevant points for each condition tested in the experiments. See \autoref{sec:protprofTools} for more options.

Region \num{4} shows a summary of the results for a selected proteins. Proteins can be selected in the listbox in Region \num{1} or in the volcano plot of Region \num{2}. The information includes a summary of the selected protein including the number of the protein in the listbox, the gene name and the protein id. Calculated p and $log_2[FC]$ values as well as averages and standard deviations for intensities and ratios.

\subsection{The Tools menu}
\label{sec:protprofTools}

The Tools menu for the results window of the Proteome Profiling module allows to further customize the plots and to apply different filters to the protein list shown in the window.

Under the menu entry Volcano Plot, users can change the condition and relevant point shown in the volcano plot, the Z score value used to color the points in the plot and the $\alpha$ value. An image of the volcano plot can also be created. If the condition or relevant point displayed is changed the Apply Filters menu entry allows to recalculate the filters for the current condition, relevant point displayed.

Under the menu entry Relevant Points, users can show all the proteins at once with different or the same colors and create an image of the Relevant Points graph.

The menu entry Export Data can be used to export the data shown in the window to a plain text file (see \autoref{subsec:utilExpData}) while the Corrected P values entry will display the information in the .protprof file using the corrected p values instead of the regular p values.

\subsubsection{Filters}

The menu entry Filters allows to Add or Remove the filters applied to the protein list. The idea behind Filters is to identify proteins with a desired behavior and discard the rest of the proteins from the listbox in Region \num{1} and the plots in Regions \num{2} and \num{3}. Filters are applied to the current Condition and Relevant point shown in the Volcano plot in Region \num{2}. If the Condition or the Relevant point shown is changed, the new plot will show the proteins obtained after the filter was applied. This allows to follow the behavior of the filtered proteins in all Conditions and Relevant points. The menu entry Applied Filters in the Volcano plot submenu allows to recalculate the filters based on the new Conditions and Relevant point shown. Any number of filters can be applied. The applied filters are shown in the bottom left corner of the results window. 

Filter can be removed in any given order using the menu entry Any in the Remove filter submenu. Additionally, the last applied filter can be removed with the menu entry Last Added or the shortcut Ctrl/Cmd + Z.

Currently, the implemented filters are:

\textbf{\textit{\underline{Z score}}}

This menu entry allows to filter proteins by the Z score value of the Fold change. 

\textbf{\textit{\underline{Log2FC}}}

This menu entry allows to filter proteins by the absolute value of the $log_2[FC]$. 

\textbf{\textit{\underline{P value}}}

This menu entry allows to filter proteins by the p value calculated for the comparison of the currently displayed  Condition and Relevant point to the control experiment. The threshold p value can be given in the 0 to 1 range or as a $-log10$ value. Regular or corrected P values can be used in the filter.

\textbf{\textit{\underline{$\alpha$ value}}}

This menu entry allows to filter proteins by the p values calculated for the comparison of the relevant points. The returned list of proteins consists of proteins for which the calculated p value is less than the selected $\alpha$ value, for at least one relevant point.

\textbf{\textit{\underline{Monotonic}}}

This menu entry allows to filter proteins by the behavior of the $log_2[FC]$ along the Relevant points. The filter searches for proteins that have a monotonically increasing or decreasing (or both) behavior for the condition shown in the Volcano plot. 

\textbf{\textit{\underline{Divergent}}}

This menu entry allows to filter proteins by the behavior of the $log_2[FC]$ along the Relevant points. In this case, the filter searches for proteins that have a monotonically increasing and decreasing behavior in at least two of the conditions tested.




































