\chapter{Utilities}
\label{chap:util}

Currently, there are \num{15} utilities. Users can access the utilities in two ways. From the main interface (\autoref{fig:mainWindow}) users can select Utilities in the list to the right and a new window will appear with a complete list of available utilities (\autoref{fig:utilWindow}). The alternative option is to directly select the desired utility from the menu entry, Utilities. The second approach is faster since does not require to use the Utilities window.

The utilities are organized in General Utilities and utilities that are specific for a given module. The following sections describe each one of the implemented utilities.

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/UTIL-WINDOW/util.jpg}	    
	\caption[The Utilities window]{\textbf{The Utilities window.} From this window users can access all the available utilities.} 
	\label{fig:utilWindow}
	\vspace{-5pt} 	
\end{figure} 

\section{Limited Proteolysis utilities}

\subsection{Sequence highlight}
\label{subsec:utilSeqHigh}
The Sequence Highlight utility allows user to highlight on the sequence of the Target protein the peptides detected in each gel spot considered in the .limprot file. The sequences with the highlighted peptides are saved in a .pdf file. How to generate a .limprot file is discussed in \autoref{chap:limprot}.

\textit{\textbf{The interface}}

The window of the Sequence Highlight utility is divided in three regions. 

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/UTIL-SEQUENCE-HIGHLIGHT/sequence-highlight.jpg}	    
	\caption[The Sequence Highlight utility window]{\textbf{The Sequence Highlight utility window.} This window allows to generate a .pdf file showing the location of the peptides detected in the investigated gel spot on the sequence of the Target protein.} 
	\label{fig:seqhighwindow}
	\vspace{-5pt} 	
\end{figure} 

Region \num{1} contains three buttons allowing users to quickly delete all provided input to generate a new .pdf file. The Clear all button will delete all user provided input. The Clear files button will delete all values in section Files of Region \num{2}. Finally, the Clear values button will delete all values in section Values of Region \num{2}.

Region \num{2} contains the fields where users provide the information needed in order to create the .pdf file with the highlighted sequences. The Limprot file button allows users to browse the file system to select the .limprot file that will be used for the generation of the .pdf file. Only one .limprot file can be provided here. The Output file button allows users to browse the file system to select the location and name of the .pdf file. If left empty, then the .pdf file, resulting from the analysis, will be saved in the same directory containing the .limprot file and will have the same name as the .limprot file. If the folder containing the selected .limprot file already contains a .pdf file with the same name as the selected .limprot file, then UMSAP will add the current date and time to the second to the end of the .pdf file name in order to avoid overwriting the older .pdf file without explicit user permission.

The Sequence length parameter allows user to define the maximum number of residues per line to be used during the creation of the .pdf file. The value here must be an integer number greater than zero.

Region \num{3} contains the Help and Start analysis buttons and a progress bar. The Help button leads to an online tutorial while the Start analysis button will start the generation of the .pdf file. The progress bar will give users a rough idea of the remaining processing time once the analysis is started.

\textit{\textbf{The analysis}}

First, UMSAP will check the validity of the user provided input. After this, for each gel spot analyzed in the .limprot file, UMSAP will print the sequence of the Target protein to the .pdf file with the peptides found in the gel spot highlighted in red.

\textit{\textbf{The output}} 

The output is a .pdf file containing a page for each gel spot found in the .limprot file. As described before, each page contains the sequence of the Target protein with the detected peptides highlighted in red. In addition, the residue number of the beginning and ending of the highlighted fragments are also given. If the sequence of the native protein was provided when creating the .limprot file, then the information is given for the recombinant and native protein. 

\textit{\textbf{The Tools menu}}

This utility does not have a Tool menu.

\section{Targeted Proteolysis utilities }  

\subsection{AA Distribution}

\subsection{Cleavages per Residue}


\subsection{Cleavages to PDB Files}

\subsection{Histograms}

\subsection{Sequence Alignments}
\label{subsec:utilSeqAli}
The Sequence Alignments utility generates sequence alignments between the FP for each experiment and the sequence of the recombinant protein. The list of FP is generated from a .tarprot file, see page \pageref{par:tarprotPIP}. The list of FP is a non redundant list. How to generate the .tarprot file is discussed in \autoref{chap:tarprot}.

\textit{\textbf{The interface}}

The Sequence Alignments window is divided in three regions, see \autoref{fig:utilSeqAli}.

Region \num{1} contains three buttons allowing users to quickly delete all provided input to generate a new .pdf file. The Clear all button will delete all user provided input. The Clear files button will delete all values in section Files of Region \num{2}. Finally, the Clear values button will delete all values in section Values of Region \num{2}.

Region \num{2} contains the fields where users provide the information needed in order to generate the alignments. The Tarprot file button allows users to browse the file system to select the .tarprot file that will be used for the analysis. Only one .tarprot file can be provided here. The Sequence alignments utility generates multiple files that will be saved in a folder named Sequences. The Output folder button allows users to browse the file system to select a location for the output folder Sequences. If the Output folder option is left empty, the output folder Sequences will be created in the same directory as the .tarprot file. If there is a Sequence folder in the selected Output folder, then UMSAP will create a new Sequences folder with the date and time to the second added to the end of the name in order to avoid overwriting any file. 

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/UTIL-SEQ-WINDOW/util-seq.jpg}	    
	\caption[The Sequence Alignments utility window]{\textbf{The Sequence Alignments utility window.} This window allows to generate sequence alignment files between the FP of each experiment and the sequence of the recombinant protein under study.} 
	\label{fig:utilSeqAli}
	\vspace{-5pt} 	
\end{figure} 

The parameter Sequence length allows to define the maximum number of residues per line in the short version of the sequence alignment files. The value here is expected to be an integer greater than zero.

Region \num{3} contains the Help and Start analysis buttons and a progress bar. The Help button leads to an online tutorial while the Start analysis button will start the analysis. The progress bar will give users a rough idea of the remaining processing time once the analysis is started.

\textit{\textbf{The analysis}}

First, UMSAP will check the validity of the user provided input. After this, the list of FP will be generated from the .tarprot file and UMSAP will generate the sequence alignments.

\textit{\textbf{The output}}

The output of the Sequence alignments utility is composed of several files that will be saved inside a folder named Sequences. Each file will be a plain text file containing an alignment. Alignments will be generated for each experiment and for the entire FP list. The sequences of the FP in each file will be N-terminally organized. Files for the recombinant and native sequences are generated. In addition, files containing one sequence per line or the specified maximum number of residues per line are also created. The sequence alignment files can be viewed with any text editor since they are just plain text files.

\subsection{Update Results}
\label{subsec:utilUpdateRes}

As discussed in \autoref{sec:backwardCompatibility}, UMSAP can read the .tarprot file from previous versions. The Update Results utility offers a way to quickly generate files for the optional analyses allowed in the Targeted Proteolysis module that are compatible with the current version of UMSAP.

\textit{\textbf{The interface}}

The Update Results utility does not have a window since there are no options to specify. When the utility is selected users will be asked to select a .tarprot file and then users must select the output folder. That is all.

\textit{\textbf{The analysis}}

UMSAP will read the .tarprot file and will perform the optional analysis specified in the .tarprot file. This will result in the creation of up to date files for the optional analyses specified in the .tarprot file. The up to date files can be viewed with the current version of UMSAP.

\textit{\textbf{The output}}

UMSAP will generate the files discussed in this chapter as required by the specified optional analyses found in the given .tarprot file. All generated files will be saved in a TarProt-Update folder created inside the specified Output folder. If there is already a TarProt-Update folder in the Output folder, then the current date and time to the second will be add to the folder name in order to avoid overwriting previous files. 

\subsection{Custom Update of Results }
\label{subsec:utilReanalyzeTarprot}

The Custom Update of Results utility is similar to the Update Results utility because they both allows to use  a .tarprot file from an older version of UMSAP to generate files for the optional analyses available in the Targeted Proteolysis module that are compatible with the current version of UMSAP. The main difference is that with Custom Update of Results a custom update can be done.

\textit{\textbf{The interface}}

The Custom Update of Results utility does not have a window. When the utility is selected users will be asked to select a .tarprot file and then the interface for the Targeted Proteolysis module is created and the information found in the selected .tarprot file is used to fill the fields in the interface of the module, see \autoref{fig:tarprotMainWindow}.

\textit{\textbf{The analysis}}

After the interface for the Targeted Proteolysis module is created and filled with the information found in the selected .tarprot file, users may modify the values or add information to perform optional analyses that were not performed with the previous versions of UMSAP, see \autoref{chap:tarprot} for more details.

\textit{\textbf{The output}}    

The output generated depends on the options given to the Targeted Proteolysis module, see \autoref{chap:tarprot} for details.

\section{General Utilities}

\subsection{Correlation Analysis}

The Correlation Analysis utility calculates the correlation in the MS data used as input for UMSAP.

\textit{\textbf{The interface}}

The Correlation Analysis window is divided in four regions, \autoref{fig:utilCorrAnalysis}. 

Region \num{1} contains three buttons allowing users to quickly delete all provided input and start a new calculation. The Clear all button will delete all user provided input and will reset the state of the list boxes of Region \num{3}. The Clear files button will delete all values in section Files of Region \num{2} and will reset the state of the list boxes of Region \num{3}. Finally, the Clear values button will delete all values in section Values of Region \num{2}.

Region \num{2} contains the fields where users provide the information needed in order to calculate the correlation between the data. Section Files contains two buttons.

\num{1}.- The Data file button allows users to browse the file system to select the Data file that will be used for the analysis. The Data file is expected to be a plain text file with tab separated columns and the name of the columns in the first row of the file. In addition, columns to be analyzed must contain only numbers and must be of the same length. Only .txt files can be provided here. 

\num{2}.- The Output file button allows users to browse the file system to select the location and name of the output file. If left empty, the name of the output file will be the same as the Data file and the .corr file will be saved in the same folder containing the Data file. If this default behavior leads to an old file been overwritten, then the date and time to the second will be used to make the name of the .corr file unique and avoid overwriting older files. 

Section Values contains two parameters.

\num{1}.- The parameter Data normalization allows users to select a normalization algorithm to be performed before the correlation analysis. Currently, only a $Log_{2}$ normalization or no normalization is possible. The list will we expanded in future versions. 

\num{2}.- The parameter Correlation method allows to select the correlation method to use.

Region \num{3} contains two list boxes and a button. The list box to the left will display the names of the columns present in the Data file, once the Data file is selected. Loading of the column names is automatically done after selecting the Data file using the Data file button in Region \num{2} or pressing the Enter key while the text box has the focus of the keyboard. Columns in the left list box can not be deleted, except in the case of loading a different Data file or using the Clear input or Clear all buttons in Region \num{1}. 

The Add columns button in the middle of Region \num{3} will add the selected columns in the left list box to the right list box. The columns will be added to the right list box in the same order as they are selected from the left list box. 

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/UTIL-CORR-WINDOW/util-corr.jpg}	    
	\caption[The Correlation Analysis utility window]{\textbf{The Correlation Analysis utility window.} This windows allows to perform a correlation analysis of the data contained in a given Data file.} 
	\label{fig:utilCorrAnalysis}
	\vspace{-5pt} 	
\end{figure}

The list box to the right of Region \num{3} contains the columns for which the correlation analysis will be performed. Correlation between all columns in the right list box will be performed. The  order of the rows and columns in the resulting matrix containing the correlation coefficients will be the same as the order of the columns shown in the right list box. Therefore, users are advised to fill the right list box in such a way that replicates of the same experiment are consecutive to each other in the right list box. Columns in the right list box can be deleted by selecting the columns and then using the right mouse button over the right list box or using the Tools menu. Columns in the right list box will be unique, meaning that a column can only be added once. 

Region \num{4} contains the Help and Start analysis buttons and a progress bar. The Help button leads to an online tutorial while the Start analysis button will start the analysis. The progress bar will give users a rough idea of the remaining processing time once the analysis is started.

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/UTIL-CORR-WINDOW/util-corr-res.jpg}	    
	\caption[The Correlation Analysis result window]{\textbf{The Correlation Analysis result window.} The correlation coefficients are shown in a color coded matrix. Values between \numrange{-1}{0} are shown in shades of red, \num{0} is shown in white and values between \numrange{0}{1} in shades of blue. NA values are shown in green.}
	\label{fig:utilCorrAnalysisRes}
	\vspace{-5pt} 	
\end{figure} 

\textit{\textbf{The analysis}}

First, UMSAP will check the validity of the user provided input. Then, columns in the right list box are read from the Data file. The columns must contain only numbers and the same amount of rows must be found in all columns. Failing to comply with this will result in the program aborting the analysis. After this, the selected normalization procedure is applied to the data. Finally, the correlation coefficients are calculated using the selected method. If any of the coefficients cannot be calculated, then the corresponding coefficient is set to NA. After the analysis is done the results will be automatically loaded and displayed in a new window, see \autoref{fig:utilCorrAnalysisRes}. 

\textit{\textbf{The output}}

The extension .corr is reserved for a file containing the output from a correlation analysis. The results in a .corr file will be shown as a color coded matrix, see \autoref{fig:utilCorrAnalysisRes}. Values between \numrange{-1}{0} will be shown in shades of red, \num{0} will be shown as white and values between \numrange{0}{1} will be shown in shades of blue. NA values will be shown in green. The columns and rows of the matrix are the column names used to calculate the correlation. Information about a specific matrix element can be obtain by simply putting the mouse pointer over the matrix element.  

\textit{\textbf{The Tools menu}}

The Tools menu in the configuration window of the correlation analysis, see \autoref{fig:utilCorrAnalysis} allows users to empty the right list box or to remove only the selected rows.

The Tools menu in the window showing the results in a .corr file, see \autoref{fig:utilCorrAnalysisRes} allows user to create an image of the plot and to export the results to one of the modules in UMSAP. After selecting to export the data, the same window used to configure the Results - Control Experiment for the selected module will appear allowing users to configure this parameter and send the configuration to the module window. If the Data file used to generate the .corr file cannot be located in the file system, then UMSAP will ask for the location of the Data file before showing the Result - Control Experiment window. The exported information includes the path to the Data file. In addition, the data shown in the window can also be exported to a plain text file. 

\subsection{Create Input File}
\label{subsec:utilUscrFile}

The Create Input File utility allows user to read a .limprot, .protprof or .tarprot file to create a .uscr file. How to generate the .limprot, .protprof or .tarprot file is discussed in \autoref{chap:limprot}, \autoref{chap:protprof} and \autoref{chap:tarprot} respectively. The .uscr file is used to prepare a module to perform an analysis without users having to type in all the information required by the module. Thus, if a second analysis of a Data file is needed users can quickly load the .uscr file into UMSAP, apply the required modifications in the window of the module and start the analysis without having to type in or modify the options that will be the same between the old and new analysis. Each .uscr file can contain information for configuring one module and one analysis.

\textit{\textbf{The interface}}

The Create Input File utility does not have a window since there are no options to specify. When the utility is selected users will be asked to select the .limprot, .protprof or .tarprot file and then users must select the output file. That is all.

\textit{\textbf{The analysis}}

First, UMSAP will check the validity of the user provided input. Then the .limprot, .protprof or .tarprot file will be read in and the configuration values will be extracted and saved in the .uscr file. The utility is able to process .tarprot files from previous versions of UMSAP.

\textit{\textbf{The output}}

The .uscr file has a simple format in which each line has a keyword and an argument. The keyword and the argument are separated by a colon (:). The following are examples of the format of the .uscr file for ech module.\newline

\textit{Example for the Limited Proteolysis module:}

Module: Limited Proteolysis\newline
Data file: /Users/kenny/data-kbr.txt\newline
Sequence (rec): /Users/kenny/seqA.txt\newline
Sequence (nat): /Users/kenny/seqA-nat.txt\newline
Output folder: /Users/kenny/test\newline
Output name: myLimTest\newline
Target protein: Mis18alpha\newline
Score value: 10\newline
Sequence length: 100\newline
d-value: NA\newline
dm-value: 8\newline
Data normalization: Log2\newline
a-value: 0.050\newline
b-value: Equal alpha\newline
y-value: 0.8\newline
Sequence: 0\newline
Detected proteins: 34\newline
Score: 42\newline
Columns to extract: 0 1 2 3 4-10\newline
Results: 69-71; 81-83, 78-80, 75-77, 72-74, NA; NA, NA, NA, 66-68, NA; 63-65, 105-107, 102-104, 99-101, NA; 93-95, 90-92, 87-89, 84-86, 60-62\newline

\textit{Example for the Proteome Profiling module:}

Module: Proteome Profiling\newline
Data file: /Users/kenny/proteinGroups-kbr.txt\newline
Output folder: /Users/kenny/test\newline
Output name: myProtTest\newline
Score value: 320\newline
Z score: 10\newline
Data normalization: Log2\newline
a-value: 0.050\newline
Median correction: True\newline
P correction: Benjamini - Hochberg\newline
Detected proteins: 0\newline
Gene names: 6\newline
Score: 39\newline
Exclude proteins: 171 172 173\newline
Columns to extract: 0 1 2 3 4-10\newline
Results: 105 115 125, 130 131 132; 106 116 126, 101 111 121; 108 118 128, 103 113 123\newline
Conditions: DMSO, H2O\newline
Relevant Points: 30min, 1D\newline
Control Type: One Control per Column\newline
Control Label: MyControl\newline

\textit{Example for the Targeted Proteolysis module:}

Module: Targeted Proteolysis\newline
Data file: /Users/kenny/data-ms.txt\newline
Sequence (rec): /Users/kenny/data-seq.txt\newline
Sequence (nat): P31545\newline
PDB file: NA\newline
Output folder: /Users/kenny/test\newline
Output name: myTarTest\newline
Target protein: efeB\newline
Score value: 200\newline
Data normalization: Log2\newline
a-value: 0.050\newline
Positions: 5\newline
Sequence length: 100\newline
Histogram windows: 50\newline
PDB ID: 2y4f;A\newline
Sequence: 100\newline
Detected proteins: 38\newline
Score: 44\newline
Columns to extract: 0 1 2 3 4-10\newline
Results: 98-105; 109-111; 112 113 114; 115-117 120\newline

The menu entry Run Input File in the Script menu allows to load the .uscr file into UMSAP.

\subsection{Export Data}
\label{subsec:utilExpData}

The Export Data utility will read an UMSAP file and will export the data present in the file to a table in a plain text file with tab separated columns. This utility is provided just in case users need to perform further data processing on the results of UMSAP. There is no interface for this utility since the only needed information is the location of the UMSAP file and the location for the output file.  

\textit{\textbf{The analysis}}

First, UMSAP will check the validity of the user provided input. Then, the UMSAP file will be read in and if needed the column names for the data will be updated. After this, the output file is generated.

\textit{\textbf{The output}}

The output file will be a plain text file with .txt extension containing a table with the data present in the selected UMSAP file. Columns in the file will be tab separated and the first row or rows will contain the names of the columns. The following is a general description of the content of the plain text files.

\textit{\textbf{Limited proteolysis}}

The generated file contains information about each one of the detected FP. The information includes the N and C-terminus number for the recombinant and native protein sequence, The results of the equivalence test for each analyzed gel spot and the sequence and score of the peptides. The results of the equivalence test are given as 0 (not equivalent) or 1 (equivalent).

\textit{\textbf{Proteome Profiling}}

The generated file contains information about each one of the proteins present in the .protprof file selected. The information includes the protein ID, gene name and score. After these three columns the table in the file is divided in information about the volcano plots (v) and the relevant points (rp).

The information for the volcano plots includes the regular (P) and corrected (Pc) p values as calculated and as $-log$ values (pP and pPc), the fold change (log2FC and FC) as well as the average and standard deviation of the intensity values for the control and the experiment. This information is given for each condition and relevant point.

The information for the relevant points comparison includes the calculated (P) and corrected (Pc) p values for the comparison of the relevant points across the different conditions and the average and standard deviation for the intensity ratios. This information is given for each relevant point.

\textit{\textbf{Targeted proteolysis}} 

The generated file contains information about each one of the detected FP. The information includes the N and C-terminus number for the recombinant and native protein sequence, The results of the ANCOVA test for each analyzed experiment and the sequence and score of the peptides. The results of the ANCOVA test are given as 0 (not different to the control) or 1 (different to the control and slope greater than 0).

\textit{\textbf{AA distribution}}

The generated file contains a matrix with the number of times each AA appears at one of the analyzed positions. This information is given for each experiment, for the FP list as a whole and for the random cleavages (RD) generated for the $\chi^2$ test. The experiment and AA in each row is given in the first and third column of the table. Columns fourth to the end of the table contains the data with the position around the cleavages indicated in the column name.

\textit{\textbf{Cleavages per residue}}

The generated file contains the number of times each residue is found at the C-terminus of a cleavage. The residue numbers are given for the recombinant and native protein. The number of cleavages are given for each experiment and sequence in the form of absolute and normalized (0 to 1 range) values.

\textit{\textbf{Correlation analysis exported data}} 

The generated file contains a square matrix with the correlation coefficients. The name of the columns in the matrix are explicitly written in the file. The name of the rows are the same as the columns with the same order and are not written to the file.

\textbf{\textit{Histograms}}

The generated file contains the window definitions and the results for the FP and each experiment. This information is given for the recombinant and native protein sequence and considering unique cleavages or all cleavages.

\subsection{Merge aadist Files}
\label{subsec:utilMergeAadistFiles}

The .aadist files contain an AA distribution analysis as described in \autoref{subsec:utilAadistCalc}. The Merge aadist Files utility allows to merge several .aadist files into a single file.

\textit{\textbf{The interface}}

The Merge aadist Files window is divided in four regions, see \autoref{fig:utilMergeAadist}.

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/UTIL-Maadist-WINDOW/util-maadist.jpg}	    
	\caption[The Merge .aadist Files utility window]{\textbf{The Merge .aadist Files utility window.} This window allows users to merge several .aadist files in a single file.}
	\label{fig:utilMergeAadist}
	\vspace{-5pt} 	
\end{figure}

Region \num{1} contains three buttons allowing users to quickly delete all provided input to generate a new .pdf file. The Clear all button will delete all user provided input. The Clear files button will delete all values in section Files of Region \num{2}.

Region \num{2} contains the fields where users provide the information needed in order to merge the .aadist files. The aadist files button allows users to browse the file system to select multiple .aadist file from a folder. Only .aadist files can be selected here. Once the files are selected the complete path to the files will be displayed in the list box in Region \num{3}. The Output file button allows users to browse the file system to select the location and name of the output file.

Region \num{3} contains a list box showing all selected .aadist files. Files can only be added one time to the list box. The order of the files in the list box is meaningless. Selected files can be deleted from the list box by pressing the right mouse button over the list box or using the Tools menu. 

Region \num{4} contains the Help and Start analysis buttons and a progress bar. The Help button leads to an online tutorial while the Start analysis button will start the analysis. The progress bar will give users a rough idea of the remaining processing time once the analysis is started.

\textit{\textbf{The analysis}}

First, UMSAP will check the validity of the user provided input. Then, the number of positions and experiments in each .aadist files are checked. If they do not match, the merging is aborted. Finally, UMSAP check that all AA distributions were originated from the same sequence and abort the task at hand if they do not. After this, files are merged. For merging the files, UMSAP adds the number of times each amino acids appears in a given position for each file. Finally, UMSAP performs the $\chi^2$ test, as indicated in \autoref{subsec:utilAadistCalc}. The significance level for the merged file is the highest value found in all files that were merged. When all files have been merged the results will be automatically loaded and displayed in a new window, see \autoref{fig:utilAadistShow}. 

\textit{\textbf{The output}} 

The Merge aadist Files utility generates a .aadist file. This file can be visualized in the same way as the output from the AA distribution utility, see \autoref{subsec:utilAadistCalc} for more details.

\subsection{Read Output File}
\label{subsec:utilReadOutF}

The Read Output File utility simply loads an output file generated by UMSAP. After selecting this option from the Utility window (\autoref{fig:utilWindow}) or the Utilities menu entry, a dialog box will be presented allowing users to select some of the output files generated by UMSAP. Currently, only .aadist, .corr, .cutprop, .hist, .limprot, .protprof, and .tarprot files can be selected. After selecting the file, the appropriate window showing the graphical representation of the file will be created. This utility can also be started with the shortcut Ctrl/Cmd+R.

\subsection{Run Input File}
\label{subsec:utilReadUscr}

The Run Input File utility allows to load a .uscr file and launches the appropriate module. The information found in the selected .uscr file is used to fill the interface of the module. Incorrect keywords - value pairs in the .uscr are silently ignored. This utility does not have an interface since the only thing required is users to select a .uscr file. This utility can also be started with the shortcut Ctrl/Cmd+I.

\subsection{Short Data Files}
\label{subsec:utilShortDF}

The Short Data Files utilities allows users to create short versions of the Data file used to create a .limprot, .protprof or .tarprot file. How to generate the .limprot, .protprof or .tarprot file is discussed in \autoref{chap:limprot}, \autoref{chap:protprof} and \autoref{chap:tarprot} respectively.

\textit{\textbf{The interface}}

The Short Data Files window is divided in three regions, see \autoref{fig:utilShortData}.

\begin{figure}[h]
	\centering
	\includegraphics[width=0.7\textwidth]{./IMAGES/UTIL-SHORTDF-WINDOW/util-shortdf.jpg}	    
	\caption[The Short Data File utility window]{\textbf{The Short Data File utility window.} This window allows to generate smaller versions of the Data file used to generate a .limport, .protprof or a .tarprot file.} 
	\label{fig:utilShortData}
	\vspace{-5pt} 	
\end{figure} 

Region \num{1} contains three buttons allowing users to quickly delete all provided input to generate a new .pdf file. The Clear all button will delete all user provided input. The Clear files button will delete all values in section Files of Region \num{2}. Finally, the Clear columns button will delete all values in section Column numbers of Region \num{2}.

Region \num{2} contains the fields where users provide the information needed in order to generate the short data files. The UMSAP file button allows users to browse the file system to select the .limprot, .protprof or .tarprot file that will be used for the analysis. Only one file can be provided here. The Data file button allows users to browse the file system to select the Data file that will be used for the analysis. The Short Data File utility generates multiple files that will be saved in a folder named Data. The Output folder button allows users to browse the file system to select a location for the output folder Data. If the Output folder option is left empty, the output folder Data will be created in the same directory as the UMSAP file. If there is a Data folder in the selected Output folder, then UMSAP will create a new Data folder with the date and time to the second added to the end of the name in order to avoid overwriting any file. 

The parameter Columns to extract allows to define which columns from the data file will be extracted to the short data files. The values here are expected to be integers greater than zero.

Region \num{3} contains the Help and Start analysis buttons and a progress bar. The Help button leads to an online tutorial while the Start analysis button will start the analysis. The progress bar will give users a rough idea of the remaining processing time once the analysis is started.

\textit{\textbf{The analysis}}

First, UMSAP will check the validity of the user provided input. If only an UMSAP file is given, then the Data file location will be read from the UMSAP file. After this, the short data files will be written to the Data folder.

\textit{\textbf{The output}}

For the Limited Proteolysis and Targeted Proteolysis modules the output consist of three files that will be saved in a Data folder. Assuming that the name of the Target protein is efeB, the name of the files will be:\newline 
all-columns-all-efeB-records.txt \newline
selected-columns-all-efeB-records.txt \newline
selected-columns-relevant-efeB-records.txt

These files are just shorter versions of the Data file containing only relevant information about the Target protein. They are plain text files with a tabular format. The first row contains the name of the columns and columns are tab separated. The file all-columns-all-efeB-records.txt contains the same number of columns as the Data file but only the rows for the Target protein. The file selected-columns-all-efeB-records.txt contains only rows for the Target protein and the columns specified in Columns to extract. The selected-columns-relevant-efeB-records.txt file is similar to the previous file but contains only the relevant peptides of the Target protein. 

In the case of the Proteome Profiling module only one file is generated. The file contains all the rows in the Data file but only the selected columns.

In all cases, these files can be viewed with any text editor.





